### DISCUSSION TITLE
uni- or bivariate ctm

### DISCUSSION THREAD

**Discussion by carlotta-hi** [cdriveraus/ctsem#34] [Q&A]
Hi Charles,

I want to compare two self-regulatory processes in terms of how quickly they return to their set point, but some problems arise:

1. Since they were originally measured with the same item ranging from -3 – 0 - +3, with negative deviations from zero representing one process and positive deviations from zero representing the second process, I split this item into two new variables (both ranging from 0 – 3, since I used the absolute deviations for simplicity). Now I am not sure if I can estimate both processes in a continuous time model or if two separete models are needed because of this dependent measurement.
If the latter is the case: Is it possible to compare the drift parameters between these separete contiunous time models? Do I need to convert them to discrete parameters to do this?

2. I tried running a simple continous time model for negative deviations as you can see here, the output from ctModelLatex is attached:

nf <- ctModel(type='stanct', id = "id",
                 time = "tag",
                 n.latent = 1,
                 n.manifest = 1,
                   manifestNames= 'n2dx',
                   latentNames= 'nd',
                   LAMBDA= 1,
                   DRIFT= 'a1', 
                   MANIFESTMEANS= 0,
                   MANIFESTVAR= ‘me’,
                   CINT= ’cint’,
                   DIFFUSION= ‘diff’)

Since the variable has a meaningful zero-point I did not center it. However, I have to account for stable individual differences and as far as I know, you recommended to fix the manifest means and include a continuous intercept with indvarying = T. Is this sufficient? I know that including traits is possible in omx, but I did not understand how it is integrated in stan.

3. Also, I am assuming stationarity for the process. But if I include stationary = T in the ctStanFit operatior or try to manually set t0var and t0means as stationary, I get the following error message: 
Stationary option temporarily unavailable -- reductions needed to pass all CRAN checks
How can I fix this?

4. Another error message I get is when I try to run ctKalman in any form:
Warning in min(x) : no non-missing arguments to min; returning Inf
Warning in max(x) : no non-missing arguments to max; returning -Inf
Error in seq.default(timerange[1], timerange[2], timestep) : 
'from' must be a finite number

I assume that the error indicates that my model is not properly specified, but I have no idea what might be the problem. Could you explain why this error occures?

Thank you very much
Carlotta
[ctsemTex1662538174.15687.pdf](https://github.com/cdriveraus/ctsem/files/9504090/ctsemTex1662538174.15687.pdf)

**Comment by cdriveraus** [comment]
Hi Carlotta, 
1. You can estimate whatever you like as a continuous time model, the question is really whether it makes sense for your research questions and data :) The split processes you describe are pretty strange and hard to deal with I think, as you have highly non-ignorable missingness -- what are your assumptions about process 1 when you have observations for process 2? What do you hope to learn about the processes?

2. In the modern ctsem formulation using type='stanct', 'traits' are included by default as random individual variation on intercept parameters (i.e. manifestmeans and cint), and they covary with the individually varying latent process starting points given by t0means. All other parameters (e.g. drift / diffusion etc) are not individually varying by default, but can be set so either by modifying the `mymodel$pars` object directly, or by specifying like 'a1||TRUE' for variation on the drift parameter. You can see from the first line of latex output that there are two randomly varying subject parameters, with covariance. 

3. Stationarity proved quite a maintenance burden and posed problems with CRAN due to model size unfortunately, I may add it back in at some point, it can be done manually but with quite some complexity. In general, there is usually little to be gained from the assumption and much to lose -- freely estimating the start points and start covariance is my preferred approach these days.

4. I suspect there may be some problem with the time coding in your data, or you have subjects with only 1 data point? If you can't find the issue in your data you can avoid the problem by specifying the timerange argument manually. 

Good luck!

**Reply by carlotta-hi** [reply]
Hi Charles,
thank you for the answers!
I wil try to set the indvarying parameters by model$pars and take a look at my data again to find the problem with ctKalman.

Regarding 1.:
I'm not happy with this solution, even though i assume that from a theoretical point of view either one or the other process takes place, but not both at the same time. The negative deviations represent the feeling of not enough closeness to a romantic partner, the positive deviations represent the feeling of too much closeness to this partner. I thought about splitting the item because I want to compare the processes in how fast they return to equilibrium and I assume that the feeling of too much closeness is easier to regulate and thus returns faster to equilibrium than feelings of not enough closeness. 
I read about oscillating ctm, but as far as i understood, if I would use the original item from -3 to +3 to estimate both processes with it, this model would not estimate separate drift parameters for the negative deviation from equilibrium vs. the positive deviation from equilibrium. Or is there an option to compare the regulation of negative and positive deviations from equilibrium in oscillating models? Another problem is, that I do not assume that the whole process cyles from positive to negative to positive etc. Sometimes you feel not enough closeness right after you successfully regulated the feeling of not enough closeness. As I understood, this contradicts a basic assumption of these models.

Best
Carlotta

**Reply by cdriveraus** [reply]
Ok I think I see. But I also think there's a very fundamental problem. What happens if you observe: 1, 3, -2, 2, -3, 1?  If you remove the negative elements then it looks like you have a slow recovery to baseline, ie 1,3,2,1, but that is incorrect. The best way I can see to handle your question of a differential return to equilibrium conditional on whether the process is positive or negative is to keep it as one process, but make the auto effect dependent on the value of the state of the process. This is a little more complicated, you can see an example doing this kind of thing in the manual (i.e. CRAN vignette), but basically instead of just a simple character string to represent a free parameter in the diagonal of the drift matrix, you would need something like `'-log1p_exp(driftbase + driftbystate * closeness)' `, where closeness is the name of your latent process, and driftbase and driftbystate are defined as additional free parameters by including the argument: 
```
PARS=c('driftbase','driftbystate')
```
in the ctModel call. the transformation around the parameters takes them from a linear, unconstrained space (i.e. where the total of the 2 parameters and state multiplication can take on any value, positive or negative), and ensures that the resulting drift parameter can only be negative, so that it is never 'explosive', ie a positive change leads to more and more gains out to infinity. 

I would certainly get a simple linear model working and make sense of that first, then see if adding in this nonlinear component improves things.

**Reply by carlotta-hi** [reply]
Hi Charles,

thank you very much for your help! 
I run both the linear and the nonlinear model and the nonlinear model showed a much better fit, but I am not sure if the estimated values are plausible because they are larger than in the linear model and also than in the continuous time models I have seen so far.
Would you mind to take a look at the model estimation?
[fit_nonlinear.pdf](https://github.com/cdriveraus/ctsem/files/9888071/fit_nonlinear.pdf)

Best
Carlotta

**Reply by cdriveraus** [reply]
whether they are plausible depends on your data / time scale choices, in general this is much easier to see if you also look at (so maybe post here) the expectations generated by the model, as well as the model specification and not just the fitted values:

ctStanDiscretePars(myfit,plot=T)
ctKalman(myfit,plot=T)
ctKalman(myfit,kalmanvec=c('etaprior'),plot=T)
ctModelLatex(mymodel)

**Reply by carlotta-hi** [reply]
Hi Charles,

regarding model specification:
the t0var is fixed to zero (1e-3) due to identification, because individual varying t0means are specified (acutally, I am a little bit confused about this point, because the model with individual varying t0means shows a much better fit than the model with t0mean without individual variation, but doesn't t0var = 0 also lead to t0means without individual variation?)

Is there any command to remove the legend in the ctKalman output? I plotted only 10 of 240 subjects because otherwise the legend takes up so much space that the graphes are barely visible.

![ctKalman](https://user-images.githubusercontent.com/113014351/199180462-6fba4a93-d5b1-4c07-a7d2-f6ccd249bd6f.png)
![ctKalman_etaprior](https://user-images.githubusercontent.com/113014351/199180465-3b204e43-1048-40b8-b758-198f85a92789.png)
![temporal_regressions](https://user-images.githubusercontent.com/113014351/199180469-5e9d5334-0554-4af3-8c0f-210294e809c7.png)
[model.pdf](https://github.com/cdriveraus/ctsem/files/9907749/model.pdf)

**Reply by cdriveraus** [reply]
I can't see any data points in that time range? do all subjects have multiple observations, what are the average time intervals between obs? you can use regular ggplot functionality to adjust the plots (the ctStanDiscretePars help file has an example, I think ctKalman does also), or you can create your own plot by using the data output when plot=FALSE. 

t0var is somewhat confusing unfortunately yes. whenever t0means are *not* indvarying, the relevant t0var rows / columns are used (e.g. if only the first t0mean is indvarying, indices above two are used). when t0means are indvarying, the relevant rows / columns are ignored and replaced internally with free parameters -- *and* additional free correlation parameters to any other random effects specified. So, to directly answer you: if indvarying is turned off for t0means, *and* variances are set to approximately zero, then there is no individual variation. If indvarying is turned off but t0var has non-zero variation, then there is individual variation that does not correlate with other individual differences. if indvarying is turned on, then there is individual variation that may correlate with other individual differences, and whatever is specified in t0var is ignored (for whichever t0means are indvarying).

**Reply by carlotta-hi** [reply]
Oh sorry, I forgot to turn removeObs = F again. 
I excluded all subjects with less than 2 timepoints and timeintervals between observations are roughly 24h.
Regarding my former issues with ctKalman: it still does not run in the default mode, I need to specify the subjects command with fit$setup$idmap[1:n,1], otherwise I get the same error as mentioned above.

Thank you for the explanation!

![ctKalman_obs](https://user-images.githubusercontent.com/113014351/199224056-00e199ca-52a8-4c0d-ab9c-5b36cfc64b45.png)

**Reply by cdriveraus** [reply]
that's a lot of overlapping lines to process :) but it looks vaguely sensible now at least! can you describe what you think the model is doing? I'm a bit worried you've copied the nonlinear section a little too exactly -- the purpose of pointing you there is just to show how to make a parameter dependent on the system state, but it looks like you've copied the oscillating structure as well, so now your state dependent parameter is not just the autoregressive effect but is the frequency of an oscillator...

**Reply by carlotta-hi** [reply]
Oh my, you are totally right!
I always thought about the course of closeness discrepancies as an oscillating one, but now I see, that this is a very strong assumption that I don’t need for my research question.
I fitted a first-order process with the parameter transformation you suggested and it works fine, thank you very much!
Because I have dyadic data I tried to account for the dependencies within the couples and found that the easiest way is to model two processes, one for each partner in the couple. It works fine, but I want to set both drift-parameters as equal, because the parameters of the processes are very similar. Now I get this error:

```
Too many indexes, expression dimensions=2, indexes found=3
error in 'model50b87b57c2e_ctsem' at line 748, column 54
-------------------------------------------------
746:             {
747:
748: DRIFT[1,1]=-log1p_exp(PARS[1,1]+PARS[2,1]*PARS[2,1][1]);
                                                      ^
749:  DRIFT[2,2]=-log1p_exp(PARS[1,1]+PARS[2,1]*PARS[2,1][2]);
 -------------------------------------------------
Error in stanc(file = file, model_code = model_code, model_name = model_name,     :
failed to parse Stan model 'ctsem' due to the above error.
```


I think it might be a technical problem, because setting the auto-effects equal works fine with “normal” driftparameters. I also ran a model with some timedependent predictors, which I also specified to have a base and state-dependent part. To set them equal for both partners of the couple is no problem.

You can find my model specifications attached.

[2p_equal.pdf](https://github.com/cdriveraus/ctsem/files/10153850/2p_equal.pdf)

**Reply by cdriveraus** [reply]
Well, a 2nd order model doesn't necessarily imply oscillations, though that's usually the context they've been used in psychology. Such a model is *more* flexible, but probably harder work to interpret and maybe unnecessary once you get into time variation! That is a weird error, seems that somewhere in the nonlinear code things get broken when you try reference a parameter called 'state' (this is used internally), so, just call it something different (not containing 'state') for now! It's fixed on the github ctsem version but you don't really need to update.

**Comment by carlotta-hi** [comment]
Hi Charles,

I fit a model again with a base and state-dependent drift parameter and a time-dependent predictor that also consists of a base and a state-dependent parameter. The drift parameter is constrained to be negative, as you showed me.
I noticed that the estimated change in certain constellations of latent state and predictor values is no longer negative/directed toward equilibrium but positive. This seems to happen because the effect of the time-dependent predictor outweighs the drift effect. Model estimation and fit seem unproblematic, and from a purely theoretical perspective, positive change can make sense. However, I am not sure if this should happen from a model-technical perspective. Would you mind taking a look on my model?

ctModel(type='stanct', id = "couple_id",
                    n.latent = 2,
                    n.manifest = 2,
                    manifestNames=c('discm', 'discf'),
                    latentNames=c("cdm", "cdf"),
                    LAMBDA=c(1, 0, 0, 1),
                    DRIFT=c('-log1p_exp(base + bystate * cdm)', 0, 0, '-log1p_exp(base + bystate * cdf)'),
                    T0VAR= "auto",
                    T0MEANS = c("t0_m", "t0_f"),
                    MANIFESTMEANS= 0,
                    MANIFESTVAR= "free",
                    CINT= c("cint_m", "cint_f"),
                    n.TDpred = 1,
                    TDpredNames = "duration_dyad_z",
                    TDPREDEFFECT = c("dbase + dbystate * cdm",  "dbase + dbystate * cdf"),
                    DIFFUSION= "auto",
                    PARS = c("base", "bystate", "dbase", "dbystate"))

![Rplot](https://github.com/cdriveraus/ctsem/assets/113014351/840be307-9777-4d6a-b13c-b43c3ed966b8)

[ctsemTex1719904867.60792.pdf](https://github.com/user-attachments/files/16064205/ctsemTex1719904867.60792.pdf)

**Comment by cdriveraus** [comment]
That looks like an equilibrium to me, can you explain where you think the problems are? Maybe you're assuming zero should be the equilibrium point, when in fact it will be determined by the auto effect and continuous intercept?

**Reply by carlotta-hi** [reply]
Yes, overall, it looks like a stable node model. I calculate the rate of change based on the estimated effects to get a better picture of what is going on. Specifically, I fix certain variables in the model equation, such as the state of the latent process or the value of the predictor, and then compare the resulting rates of change with each other. In the present data, I noticed that under certain circumstances, a rate of change directed away from equilibrium results when you combine the values of the auto effect and the time-dependent predictors. While the closeness discrepancy (the latent process) tends to weaken/move towards zero, there are values of the time-dependent predictor that cause the closeness discrepancy to increase.
I just want to make sure that this is a result that can be produced by a properly specified continuous-time model or if this is a sign that something is wrong with my analyses.

**Reply by cdriveraus** [reply]
I don't totally follow you. A time dependent predictor will generally shift the system in a direction, and not necessarily towards equilibrium -- imagine e.g. taking a dose of medication or experiencing a traumatic event. Are your td predictors reasonably conceptualised as single events? If not, you might be better off modelling them as latent states with a measurement model. If you haven't seen it this paper has out-dated code but might be conceptually useful. 
https://www.researchgate.net/publication/328221807_Understanding_the_Time_Course_of_Interventions_with_Continuous_Time_Dynamic_Models

**Reply by cdriveraus** [reply]
To more directly get at the question, I guess the answer is 'I think things are technically fine' and there is no specific bug in ctsem or anything here, but perhaps you have some confusion in your conceptualization that is worth thinking about.

**Reply by carlotta-hi** [reply]
Thank you, that already helps me! Perhaps my question was overly cautious; I'm still not entirely sure about the specific differences between ct models and conventional models. 
I have been thinking about modeling the td predictor as latent state, so thank you for the suggestion!

**Reply by cdriveraus** [reply]
This paper (soon out in psych methods) elaborates in some detail on the difference:
https://osf.io/preprints/xdf72
